source/common/network/connection_impl.cc

```diff
@@ -395,6 +396,21 @@ void ConnectionImpl::updateWriteBufferStats(uint64_t num_written, uint64_t new_s
                                            buffer_stats_->write_current_);
 }
 
+const std::string Network::ConnectionImpl::destinationAddress() {
+  if (fd_ != -1) {
+    sockaddr_storage orig_dst_addr;
+    memset(&orig_dst_addr, 0, sizeof(orig_dst_addr));
+    bool success = Utility::getOriginalDst(fd_, &orig_dst_addr);
+    if (success) {
+      return Utility::urlForTcp(
+          Utility::getAddressName(reinterpret_cast<struct sockaddr_in*>(&orig_dst_addr)),
+          Utility::getAddressPort(reinterpret_cast<struct sockaddr_in*>(&orig_dst_addr)));
+    }
+  }
+
+  return EMPTY_STRING;
```

Comment:
I see your point on the performance issue. To me this seems strange though since you could still call this on a client connection, a UDS connection, etc. I realize it would return empty string, but it still seems off. I think I would go with adding the member and returning const&. Even though we will pay for the extra getsockopt(), I don't really see CPS as a major performance benchmark currently. If it really becomes an issue, I think we could plumb it through such that you use the getsockopt before the listener redirect to populate it. But for now my vote would be to not worry about it and get the interface right.